torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,89 +65,152 @@ def reset_table(self) -> None:
+# track mutation. Each triton language function is monkey patched to
+# either detect the mutation or return a fresh scalar object.
+def identify_mutated_tensors(kernel, kwargs):
+    from triton import language as tl
     from triton.runtime.autotuner import Autotuner
 
+    # Replace tensor pointers with Proxy object. Do not replace the other
+    # inputs since tl.constexprs can be inspected by the kernel.
+    proxy_kwargs = {
+        key: (Proxy() if isinstance(value, Tensor) else value)
+        for key, value in kwargs.items()
+    }
+
     if isinstance(kernel, Autotuner):
+        if len(kernel.configs) > 0:
+            # If we are autotuning, then it doesn't matter which version gets
+            # picked for tracing purposes, so lets pick the first one
+            proxy_kwargs = {**proxy_kwargs, **kernel.configs[0].kwargs}
         kernel = kernel.fn
 
-    infos = {name: MutationInfo() for name in tensors}
-    tracker = MutationTracker(infos)
-    tracker.visit(kernel.parse())
-    return [
-        name for name, info in infos.items() if info.mutated or info.used_in_unknown
-    ]
+    default_impls: Dict[str, FunctionType] = {}
+    mutation_ops = {
+        "store",
+        "atomic_add",
+        "atomic_cas",
+        "atomic_max",
+        "atomic_min",
+        "atomic_xchg",
+    }
+    try:
+        # Monkey patch all triton language functions
+        for name, impl in inspect.getmembers(tl, inspect.isfunction):
+            default_impls[name] = impl
+
+            if name in mutation_ops:
+                # Ops that mutate the first argument tensor pointer
+
+                def fn(pointer, *args, **kwargs):
+                    assert isinstance(pointer, Proxy)
+                    pointer.mutated = True
+                    return Scalar()
+
+            elif name == "make_block_ptr":
+                # Creates indirection to base tensor pointer
+
+                def fn(base, *args, **kwargs):  # type: ignore[misc] # different conditional signatures
+                    assert isinstance(base, Proxy)
+                    return base
+
+            elif name == "inline_asm_elementwise":
+                # If there's inline asm in the kernel, anything can happen.
+                # Do not optimize
+
+                def fn(*args, **kwargs):  # type: ignore[misc] # different conditional signatures
+                    raise IdentifyMutationException("inline_asm_elementwise in kernel")
+
+            else:
+                # Any operation not listed above, do not mutate a kernel
+                # pointer. They either return some sort of scalar or
+                # for the purposes of tracing cause an uninteresting
+                # side effect.
+
+                def fn(*args, **kwargs):  # type: ignore[misc] # different conditional signatures
+                    return Scalar()
+
+            setattr(tl, name, fn)
+
+        kernel.fn(**proxy_kwargs)
```

Comment:
> Triton only allows control flow via constexpr stuff and i pass those as their real values.

This isn't true, kernels can have dynamic control flow. The simplest example is different behavior depending on `tl.program_id`:
```python
if tl.program_id(0) == 0:
    # ...
elif tl.get_program_id(0) == 1:
   # ...
```
In which case you need to be able to visit both sides of the `if`.

I think in this case you would probably raise a type error when converting `Proxy` to a boolean so I guess this is fine though (?)


> Triton team's current stance is that to do this analysis properly, we should link against Triton's MLIR library and do an out-of-tree analysis.

tbh since this is such a simple analysis you should be able to get away with a string search on the text form of the triton IR which triton already lets you access from a compiled kernel.

Reply:
Hmm, I wasn't aware of dynamic control flow. Just tried some code with it and TTIR is
```
    %15 = scf.if %14 -> (tensor<16xf32>) {
      %18 = arith.addf %9, %12 : tensor<16xf32>
      scf.yield %18 : tensor<16xf32>
    } else {
      %18 = arith.mulf %9, %12 : tensor<16xf32>
      scf.yield %18 : tensor<16xf32>
    }
```

With the current implementation, tl.program_id(0) would return Scalar(), Scalar() == 0 would also return scalar, so `if Scalar()` would be true, so we would always look at the first branch.

I can add more bail out conditions here to handle this but I wonder @peterbell10 are there more defined conditions for dynamic control flow in Triton?

Reply:
Although I haven't seen this example in the internal kernels yet, it doesn't sound that contrived. E.g., IIRC, Inductor-generated foreach kernels have this structure of multiple branches relying on `program_id` ranges.

Is it possible to harvest dependencies from both branches? Just to cover all possible paths in the "graph" from mutating ops back to inputs.

Reply:
That would require analysis over the kernel but the current approach employs trace execution. I added a bail out if we cast Scalar to bool and added a test case for it.

I think the best way to do this analysis is still over the TTIR or something lower, and we should negotiate with Phil etc to get this into to Triton.

Reply:
> I wonder @peterbell10 are there more defined conditions for dynamic control flow in Triton?

In triton, fairly arbitrary control flow is possible with `if` and `while` loops. The only restriction being that the condition must evaluate to a scalar value, but you could for example get to that scalar via a reduction from an intermediate tensor. You could also load a value from memory and branch on that.

Reply:
This is a good point, this analysis would be incorrect on inductor-generated foreach kernels.

Reply:
Yep, yesterday, I decided against continuing with this approach after learning about dynamic control flow. I was under the impression that control flow was only via `tl.constexpr`. I'm exploring the TTIR based solution, we will need something like [MLIR Python Bindings](https://mlir.llvm.org/docs/Bindings/Python/) and a whitelist TTIR ops that write to memory, and implement the analysis ourselves.

Are we ok taking a dependency on these MLIR python bindings?

Reply:
If we somehow gain access to the Triton dialect op specs, we won't need to maintain the white list? But then I'm not sure what exactly the "TTIR based solution" would be: parsing the TTIR dump that we already have or integrating our own MLIR pass (in C++, in- or out-of-tree) into the Triton MLIR machinery.

Reply:
> If we somehow gain access to the Triton dialect op specs

We won't get this easily so for the purpose of this, I'll assume we dont have it

> I'm not sure what exactly the "TTIR based solution" would be

We can `ast_to_ttir` this gives us a module that we cannot inspect. Then we call `dump()` on this and get string TTIR. We parse this with mlir python bindings and then we are free to do our own analysis.